openedx/core/djangoapps/user_api/accounts/views.py

```diff
@@ -765,7 +779,28 @@ def partial_update(self, request):
         """
         try:
             username = request.data['username']
-            retirement = UserRetirementStatus.objects.get(original_username=username)
+            retirements = UserRetirementStatus.objects.filter(original_username=username)
+
+            # During a narrow window learners were able to re-use a username that had been retired if
+            # they altered the capitalization of one or more characters. Therefore we can have more
+            # than one row returned here (due to our MySQL collation being case-insensitive), and need
+            # to disambiguate them in Python, which will respect case in the comparison.
+            retirement = None
+            if len(retirements) < 1:
+                raise UserRetirementStatus.DoesNotExist()
+            elif len(retirements) > 1:
+                for r in retirements:
+                    if r.original_username == username:
+                        retirement = r
+                        break
+                # UserRetirementStatus was found, but it was the wrong case.
+                if retirement is None:
+                    raise UserRetirementStatus.DoesNotExist()
+            else:
+                if username != retirements[0].original_username:
+                    raise UserRetirementStatus.DoesNotExist()
+                retirement = retirements[0]
```

Comment:
nit: can this else case be folded into the >1 case by just removing lines 799-802 and changing the condition to >= 1?